src/hooks/useBasisForm/useBasisField.ts

```diff
@@ -0,0 +1,113 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import {
+  FieldPathValue,
+  FieldPath,
+  FieldValues,
+  useController,
+  useFormContext,
+  RegisterOptions,
+} from "react-hook-form";
+import { ComponentName } from "../../components/componentName";
+import { getRulesFromComponent } from "./getRulesFromComponent";
+import { nameToDefaultValueMap } from "./nameToDefaultValueMap";
+import { getBasisErrors } from "./validation/getBasisErrors";
+
+interface FieldOptions<
+  TFieldValues extends FieldValues,
+  Name extends FieldPath<TFieldValues>
+> {
+  name: Name;
+  componentDisplayName?: string;
+  componentProps?: Record<string, any>;
+  customRules?: RegisterOptions<TFieldValues, Name>;
+  defaultValue?: FieldPathValue<TFieldValues, Name>;
+}
+
+export const useBasisField = <
+  TFieldValues extends FieldValues,
+  Name extends FieldPath<TFieldValues>
+>(
+  props: FieldOptions<TFieldValues, Name>
+) => {
+  const {
+    name,
+    defaultValue,
+    customRules,
+    componentProps = {},
+    componentDisplayName = "",
+  } = props;
+  const componentName = componentDisplayName as ComponentName;
+
+  const { trigger } = useFormContext<TFieldValues>();
+
+  const componentDefaultValue =
+    nameToDefaultValueMap[componentName];
+
+  const { field, fieldState } = useController({
+    name,
+    defaultValue: defaultValue ?? componentDefaultValue,
+    rules: customRules || getRulesFromComponent(componentName, componentProps),
+  });
+
+  const [hasBeenInvalid, setHasBeenInvalid] = useState(false);
+
+  useEffect(() => {
+    if (fieldState.invalid && !hasBeenInvalid) {
+      setHasBeenInvalid(true);
+    }
+  }, [fieldState.invalid, hasBeenInvalid]);
+
+  // functions in ref to avoid recreating event listeners when these change
+  const functionRefs = useRef({
+    trigger,
+    componentOnChange: componentProps.onChange,
+    componentOnBlur: componentProps.onBlur,
+    fieldOnChange: field.onChange,
+    fieldOnBlur: field.onBlur,
+  });
+
+  useEffect(() => {
+    functionRefs.current.trigger = trigger;
+    functionRefs.current.componentOnChange = componentProps.onChange;
+    functionRefs.current.componentOnBlur = componentProps.onBlur;
+    functionRefs.current.fieldOnChange = field.onChange;
+    functionRefs.current.fieldOnBlur = field.onBlur;
+  });
+
+  const onChange = useCallback(
+    (...args) => {
+      functionRefs.current.fieldOnChange(...args);
+      functionRefs.current.componentOnChange?.(...args);
+      if (hasBeenInvalid) {
+        functionRefs.current.trigger(name);
+      }
+    },
+    [name, hasBeenInvalid]
+  );
+
+  const onBlur = useCallback(
+    (...args) => {
+      functionRefs.current.componentOnBlur?.(...args);
+      functionRefs.current.fieldOnBlur();
```

Comment:
shouldn't we pass args here?

Reply:
We don't need to pass the args as the on blur of field does not require any event or argument

Reply:
I can't do that as typescript gives an error as the function does not expect any arguments. So if they do change it at any point typescript will let us know